### PR TITLE
resource/aws_appsync_resolver: Add CachingConfig support

### PR DESCRIPTION
**Pull request by ackintosh**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

>

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
ENHANCEMENTS:

* resource/aws_appsync_resolver: Add `caching_config` block
```

Output from acceptance testing:

```
AWS_PROFILE=default make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppsyncResolver_CachingConfig'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsAppsyncResolver_CachingConfig -timeout 120m
=== RUN   TestAccAwsAppsyncResolver_CachingConfig
=== PAUSE TestAccAwsAppsyncResolver_CachingConfig
=== CONT  TestAccAwsAppsyncResolver_CachingConfig
--- PASS: TestAccAwsAppsyncResolver_CachingConfig (89.42s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       89.490s
```

### CODE REVIEW

**Comment by ackintosh** [comment]
The CI job has completed successfully but the status on github is still Pending... 🤔 

https://travis-ci.org/github/terraform-providers/terraform-provider-aws/builds/673029687

**Comment by ackintosh** [comment]
I've force-pushed in order to restart the job...

**Comment by ackintosh** [comment]
Whew, this PR is ready for review!

**Comment by ackintosh** [comment]
This PR closes https://github.com/terraform-providers/terraform-provider-aws/issues/13230

File: aws/resource_aws_appsync_resolver.go [hunk 1]
```diff
@@ -77,6 +77,25 @@ func resourceAwsAppsyncResolver() *schema.Resource {
 					},
 				},
 			},
+			"caching_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
```
Comment (ewbankkit) [inline aws/resource_aws_appsync_resolver.go:83]: Please add `MaxItems: 1,` .

File: aws/resource_aws_appsync_resolver.go [hunk 2]
```diff
@@ -77,6 +77,25 @@ func resourceAwsAppsyncResolver() *schema.Resource {
...
+					Schema: map[string]*schema.Schema{
+						"caching_keys": {
+							Type:     schema.TypeList,
```
Comment (ewbankkit) [inline aws/resource_aws_appsync_resolver.go:86]: I suggest 

```hcl
Type:     schema.TypeSet,
Set:      schema.HashString,
```

here as order is not important.

File: aws/resource_aws_appsync_resolver.go [hunk 3]
```diff
@@ -243,3 +274,21 @@ func decodeAppsyncResolverID(id string) (string, string, string, error) {
...
+
+func expandAppsyncResolverCachingConfig(l []interface{}) *appsync.CachingConfig {
+	if len(l) < 1 || l[0] == nil {
+		return nil
+	}
+
+	m := l[0].(map[string]interface{})
+
+	cachingConfig := &appsync.CachingConfig{
+		CachingKeys: expandStringList(m["caching_keys"].([]interface{})),
```
Comment (ewbankkit) [inline aws/resource_aws_appsync_resolver.go:286]: `CachingKeys: expandStringSet(m["caching_keys"].(*schema.Set)),`

File: aws/structure.go [hunk 4]
```diff
@@ -5565,6 +5565,23 @@ func flattenAppsyncPipelineConfig(c *appsync.PipelineConfig) []interface{} {
...
+	if c == nil {
+		return nil
+	}
+
+	if len(c.CachingKeys) == 0 && *(c.Ttl) == 0 {
+		return nil
+	}
+
+	m := map[string]interface{}{
+		"caching_keys": flattenStringList(c.CachingKeys),
```
Comment (ewbankkit) [inline aws/structure.go:5578]: `flattenStringSet(c.CachingKeys)`

File: aws/structure.go [hunk 5]
```diff
@@ -5565,6 +5565,23 @@ func flattenAppsyncPipelineConfig(c *appsync.PipelineConfig) []interface{} {
 	return []interface{}{m}
 }
 
+func flattenAppsyncCachingConfig(c *appsync.CachingConfig) []interface{} {
```
Comment (ewbankkit) [inline aws/structure.go:5568]: Maybe move to `aws/resource_aws_appsync_resolver.go` to avoid overloading this file?

Comment (ackintosh) [reply aws/structure.go:5568]: That makes sense, thank you!

I'll also move `flattenAppsyncPipelineConfig` together with `flattenAppsyncCachingConfig ` to `aws/resource_aws_appsync_resolver.go`. 💡 

Btw, besides `flattenAppsync**`, I can see some resource-specific function as well in `structure.go` but I'm not sure why the functions are in the `structure.go`. 🤔

**Comment by ackintosh** [comment]
After the updates, I've checked the acceptance tests have passed. 😌 

Output:
```shell-session
AWS_PROFILE=default make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppsyncResolver_CachingConfig'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsAppsyncResolver_CachingConfig -timeout 120m
=== RUN   TestAccAwsAppsyncResolver_CachingConfig
=== PAUSE TestAccAwsAppsyncResolver_CachingConfig
=== CONT  TestAccAwsAppsyncResolver_CachingConfig
--- PASS: TestAccAwsAppsyncResolver_CachingConfig (66.91s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       66.963s
```

**Comment by ackintosh** [comment]
After the updates, I've checked the acceptance tests have passed. 😌 

Output:
```shell-session
$ AWS_PROFILE=default make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppsyncResolver_CachingConfig'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsAppsyncResolver_CachingConfig -timeout 120m
go: downloading github.com/aws/aws-sdk-go v1.30.21
=== RUN   TestAccAwsAppsyncResolver_CachingConfig
=== PAUSE TestAccAwsAppsyncResolver_CachingConfig
=== CONT  TestAccAwsAppsyncResolver_CachingConfig
--- PASS: TestAccAwsAppsyncResolver_CachingConfig (74.90s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       74.945s
```

**Comment by ackintosh** [comment]
@ewbankkit Thank you for your review! Please have another look when you have time. 😃

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAppsyncResolver_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAwsAppsyncResolver_ -timeout 120m
=== RUN   TestAccAwsAppsyncResolver_basic
=== PAUSE TestAccAwsAppsyncResolver_basic
=== RUN   TestAccAwsAppsyncResolver_disappears
=== PAUSE TestAccAwsAppsyncResolver_disappears
=== RUN   TestAccAwsAppsyncResolver_DataSource
=== PAUSE TestAccAwsAppsyncResolver_DataSource
=== RUN   TestAccAwsAppsyncResolver_RequestTemplate
=== PAUSE TestAccAwsAppsyncResolver_RequestTemplate
=== RUN   TestAccAwsAppsyncResolver_ResponseTemplate
=== PAUSE TestAccAwsAppsyncResolver_ResponseTemplate
=== RUN   TestAccAwsAppsyncResolver_multipleResolvers
=== PAUSE TestAccAwsAppsyncResolver_multipleResolvers
=== RUN   TestAccAwsAppsyncResolver_PipelineConfig
=== PAUSE TestAccAwsAppsyncResolver_PipelineConfig
=== RUN   TestAccAwsAppsyncResolver_CachingConfig
=== PAUSE TestAccAwsAppsyncResolver_CachingConfig
=== CONT  TestAccAwsAppsyncResolver_basic
=== CONT  TestAccAwsAppsyncResolver_multipleResolvers
=== CONT  TestAccAwsAppsyncResolver_CachingConfig
=== CONT  TestAccAwsAppsyncResolver_PipelineConfig
=== CONT  TestAccAwsAppsyncResolver_RequestTemplate
=== CONT  TestAccAwsAppsyncResolver_ResponseTemplate
=== CONT  TestAccAwsAppsyncResolver_DataSource
=== CONT  TestAccAwsAppsyncResolver_disappears
--- PASS: TestAccAwsAppsyncResolver_PipelineConfig (29.86s)
--- PASS: TestAccAwsAppsyncResolver_disappears (30.41s)
--- PASS: TestAccAwsAppsyncResolver_CachingConfig (37.19s)
--- PASS: TestAccAwsAppsyncResolver_multipleResolvers (41.31s)
--- PASS: TestAccAwsAppsyncResolver_ResponseTemplate (46.97s)
--- PASS: TestAccAwsAppsyncResolver_DataSource (50.75s)
--- PASS: TestAccAwsAppsyncResolver_RequestTemplate (50.87s)
--- PASS: TestAccAwsAppsyncResolver_basic (75.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	75.754s
```

**Pull request merged**